fs: add uv_fs_utime_ex and uv_fs_futime_ex

whitlockjc opened the pull request:
These new functions behave identicaly to uv_fs_utime and uv_fs_futime
respectively for every OS other than OS X and Windows.  For OS X and
Windows these new APIs will allow you to alter the birth/creation time
of the file as well as the access and modify times.

Fixes #499

whitlockjc commented:
I just realized that I made the signature `uv_fs_(f)utime(..., btime, atime, mtime, ...)` but with `uv_fs_(f)utime` looking like `uv_fs_(f)utime(..., atime, mtime, ...)`, I wonder if I should had made the signature like `uv_fs_(f)utime(..., atime, mtime, btime, ...)`.  Of course, I'll await review feedback.

whitlockjc commented:
Also, now that I implemented this only for OS X and Windows, I wonder if the OS X specific logic in `src/unix/fs.c` should be moved to `src/unix/darwin.c`.  Again, your decision.

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -188,9 +193,38 @@ static ssize_t uv__fs_futime(uv_fs_t* req) {
+  struct timespec times[3];
+  unsigned i = 0;
+
+  bzero(&attr_list, sizeof(attr_list));
```
Please use memset.

whitlockjc replied:
You got it.

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -188,9 +193,38 @@ static ssize_t uv__fs_futime(uv_fs_t* req) {
+    attr_list.commonattr |= ATTR_CMN_CRTIME;
+
+    times[i++].tv_sec = req->btime;
+    times[i].tv_nsec = (unsigned long)(req->btime * 1000000) % 1000000 * 1000;
```
Just curious, is there a reason you're reducing to microseconds here when it can do nanoseconds?

whitlockjc replied:
I just followed the convention I saw.  Maybe I chose the wrong one.

bnoordhuis replied:
Can you point me to where you saw that?  Or were those timeval structs?

whitlockjc replied:
https://github.com/libuv/libuv/blob/v1.x/src/unix/fs.c#L153

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -188,9 +193,38 @@ static ssize_t uv__fs_futime(uv_fs_t* req) {
+    times[i].tv_nsec = (unsigned long)(req->mtime * 1000000) % 1000000 * 1000;
+  }
+
+  return fsetattrlist(req->file, &attr_list, times, i * sizeof(times), 0);
```
Shouldn't that be `i * sizeof(times[0])`?

whitlockjc replied:
My fault, I meant for it to be `i * sizeof(timespec)`.

bnoordhuis replied:
Please use `i * sizeof(times[0])` in that case.  It doesn't really matter here (apart from being libuv convention) but in general it's better to take the size of variables rather than types because the type of a variable can change.

whitlockjc replied:
Yes, sorry about that.

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -611,10 +645,43 @@ static ssize_t uv__fs_sendfile(uv_fs_t* req) {
+
+    times[i++].tv_sec = req->mtime;
+    times[i].tv_nsec = (unsigned long)(req->mtime * 1000000) % 1000000 * 1000;
+  }
```
Too much duplication with uv_fs_futime, this should be DRYd.

whitlockjc replied:
I thought so as well but I saw a lot of the same elsewhere.  I guess this is slightly different because it's 15 lines vs. 8-10 elsewhere.

bnoordhuis commented on src/win/fs.c:
```diff
@@ -1403,13 +1404,18 @@ static void fs__fchmod(uv_fs_t* req) {
+  int btime_provided = !isnan(btime);
+
+  if (btime_provided) {
+    TIME_T_TO_FILETIME((time_t) btime, &filetime_b);
```
Does the cast to `time_t` mean there is no sub-second precision on Windows?

whitlockjc replied:
Followed convention again.  If we change this for `btime`, we'd need to do the same for `atime` and `mtime` which were already using this pattern.

bnoordhuis replied:
Oh, I wasn't saying it's wrong, just wondering out aloud.

whitlockjc replied:
Oh, I wasn't worried.  Just explaining that the code pre-existed this change.  ;)  I'm up for cleaning things up of course.

piscisaureus replied:
> Does the cast to time_t mean there is no sub-second precision on Windows?

The cast is a bit strange actually ... the unit of time measurement on NT(FS) is 100ns so we should be able to support subsecond precision.

whitlockjc replied:
I don't disagree but that would mean changing code that lived before this patch.  That said, I think that enhancement should be a separate PR.  I'd be glad to help of course.

bnoordhuis commented on src/win/fs.c:
```diff
@@ -1403,13 +1404,18 @@ static void fs__fchmod(uv_fs_t* req) {
   TIME_T_TO_FILETIME((time_t) mtime, &filetime_m);
 
-  if (!SetFileTime(handle, NULL, &filetime_a, &filetime_m)) {
+  if (!SetFileTime(handle, btime_provided ? &filetime_b : NULL, &filetime_a, &filetime_m)) {
```
Can you fix the long lines?

whitlockjc replied:
Sure.  I noticed that this file has a number of formatting issues and wasn't sure which to honor.

bnoordhuis commented:
Left some comments.  For the record: the changes to `UV_FS_PRIVATE_FIELDS` means it's a semver-major change and can't land in the v1.x branch.

whitlockjc commented:
Changing `UV_FS_PRIVATE_FIELDS` meaning a semver-major change...glad to know.  I'll fix this PR for the changes above and then I can rebase this off of the appropriate location.

bnoordhuis commented:
Just FYI, master is currently a great deal behind because we haven't started work on v2.x yet.

whitlockjc commented:
Any suggestion?

bnoordhuis commented:
We normally merge v1.x into master after a release.  I can jump the line but I speculate we'll be making a new release in a few days anyway.

whitlockjc commented:
PR updated with your requested changes.

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -58,6 +59,41 @@
+                                                                               \
+    times[i++].tv_sec = req->mtime;                                            \
+    times[i].tv_nsec = (unsigned long)(req->mtime * 1000000) % 1000000 * 1000; \
+  }
```
I'd make this a function, not a macro.  Rule of thumb: a macro when you must, a function when you can.

whitlockjc replied:
You got it.

whitlockjc replied:
Just so I can better understand this rule, there are already macros for `INIT`, `PATH`, `PATH2` and `POST` so how are they different than this new one?  I ask because I used those as the reasoning behind using a macro.  I'm completely cool with changing regardless but I wanted to better understand for next time.

bnoordhuis replied:
Those could be functions too.  IIRC in their original incarnation they had more macro magic in them; now not so much.

whitlockjc replied:
Gotcha.  Like I said, I will do it anyways.  Thanks for taking the time to help me.

whitlockjc replied:
First pass is in the latest version of the PR.

whitlockjc commented:
PR updated with your requested changes.  *(`SETATTRLIST` macro turned into `uv__prepare_setattrlist_args` function)*

saghul commented on include/uv-win.h:
```diff
@@ -609,6 +609,7 @@ RB_HEAD(uv_timer_tree_s, uv_timer_s);
       uv_buf_t bufsml[4];                                                     \
     } info;                                                                   \
     struct {                                                                  \
+      double btime;                                                           \
```
ABI break?

whitlockjc replied:
Yeah, @bnoordhuis already commented on this.  Looks like this won't land in `v1.x`.

saghul commented on docs/src/fs.rst:
```diff
@@ -258,6 +258,13 @@ API
+    are on OS X or Windows in which case these variants allow you to also set
+    the birth/creation time.
+
 .. c:function:: int uv_fs_link(uv_loop_t* loop, uv_fs_t* req, const char* path, const char* new_path, uv_fs_cb cb)
```
Needs a ".. versionadded". @bnoordhuis this adds a new double to `uv_fs_t`, not sure we can land it on v1.x.

whitlockjc replied:
Commented above.  I can update the version added stuff when we figure out when it will land.  Cool?

saghul commented:
Thanks for working on this @whitlockjc! :heart:

whitlockjc commented:
Any time.  Let me know if I can help further.  *(I'll go through the issue backlog within the next few days.)*

jorangreef commented:
Thanks @whitlockjc

jorangreef commented:
@whitlockjc, I patched Node 5.1.0 `deps/uv` with the latest patch from your pull request (https://github.com/libuv/libuv/pull/590.patch), rewriting the paths in the patch to work against `deps/uv` and manually checking that all the individual diffs came through.

I didn't add any Node binding for the new `...utime_ex` methods, I just applied the patch to `deps/uv`, compiled Node and ran `make test` to see if Node's tests would pass, but I got the following errors:

```
=== release test-fs-utimes ===
Path: parallel/test-fs-utimes
FAILED: expect_ok { '0': 'utimes',
  '1': '/Users/Joran/More/node/test/parallel/test-fs-utimes.js',
  '2': undefined,
  '3': '123456',
  '4': -1 }
FAILED: expect_ok { '0': 'futimes', '1': 15, '2': undefined, '3': '123456', '4': -1 }
FAILED: expect_ok { '0': 'utimesSync',
  '1': '/Users/Joran/More/node/test/parallel/test-fs-utimes.js',
  '2': undefined,
  '3': '123456',
  '4': -1 }
FAILED: expect_ok { '0': 'futimesSync',
  '1': 15,
  '2': undefined,
  '3': '123456',
  '4': -1 }
Tests run / ok: 48 / 44
```

Is there something I've missed that I need to update in the Node bindings for Node's existing fs.utimes to work? I noticed that libuv now delegates uv_fs_futime to uv_fs_futime_ex and that this assigns `req->btime = btime`. Could this have something to do with it?

I wanted to run the tests and then, if these pass, start work on adding the binding on the Node side in the meantime.

whitlockjc commented:
Yes.  If you see @bnoordhuis' comment [here](https://github.com/libuv/libuv/pull/590#issuecomment-151667151), you'll see that the this PR changes `UV_FS_PRIVATE_FIELDS` which means the structures returned will now have the new `btime` member and that is why this likely won't land until `2.x`.  I'm not surprised that Node.js building against this might have test failures.

As for the work to support this landing in Node.js, I'd love to do it when the time comes but if you beat me to it, no big deal.

jorangreef commented:
> As for the work to support this landing in Node.js, I'd love to do it when the time comes

Thanks, that would be great.

cjihrig commented:
@whitlockjc are you interested in picking this back up? If so, it needs a rebase and should target `master` instead of `v1.x`.

whitlockjc commented:
Yeah, I’ll pick it back up.

whitlockjc commented:
I'm working on this over the weekend.  Thanks for your patience.

whitlockjc commented:
I have rebased against `master` and made appropriate fixes.  I've tested on OS X and I will soon test on Linux and Windows.  Please re-review the changes.

whitlockjc commented:
Can we remove the `stalled` label?

cjihrig commented on test/test-fs.c:
```diff
@@ -2313,6 +2383,65 @@ TEST_IMPL(fs_futime) {
+  r = uv_fs_open(NULL, &req, path, O_RDWR, 0, NULL);
+  ASSERT(r >= 0);
+  ASSERT(req.result >= 0);
+  file = req.result; /* FIXME probably not how it's supposed to be used */
```
FWIW, in test-fs.c, we do this `fd = uv_get_osfhandle(open_req1.result);`

cjihrig replied:
Missed this one.

whitlockjc replied:
Yeah, my bad.

cjihrig commented on src/win/fs.c:
```diff
@@ -2374,9 +2400,16 @@ int uv_fs_utime(uv_loop_t* loop, uv_fs_t* req, const char* path, double atime,
   req->fs.time.mtime = mtime;
   POST;
-}
+}
```
Missing newline?

cjihrig commented on src/win/fs.c:
```diff
@@ -1497,13 +1500,24 @@ static void fs__fchmod(uv_fs_t* req) {
+                                   double btime,
+                                   double atime,
+                                   double mtime) {
+  FILETIME filetime_b, filetime_a, filetime_m;
```
Can you split this into three separate declarations.

cjihrig commented on src/win/fs.c:
```diff
@@ -1497,13 +1500,24 @@ static void fs__fchmod(uv_fs_t* req) {
+  FILETIME filetime_b, filetime_a, filetime_m;
+  int btime_provided = !isnan(btime);
+
+  if (btime_provided) {
```
Instead of having `btime_provided`, could you make the `isnan()` call the conditional, and just assign `filetime_b` to a value or `NULL`. Then, you can also get rid of the ternary below.

cjihrig commented on src/unix/fs.c:
```diff
@@ -704,10 +748,20 @@ static ssize_t uv__fs_sendfile(uv_fs_t* req) {
 static ssize_t uv__fs_utime(uv_fs_t* req) {
+#if defined(__APPLE__)
+  struct attrlist attr_list;
+  unsigned i = 0;
```
Can you separate the declaration from the assignment.

cjihrig commented on src/unix/fs.c:
```diff
@@ -230,9 +268,15 @@ static ssize_t uv__fs_futime(uv_fs_t* req) {
   return r;
+#elif defined(__APPLE__)
+  struct attrlist attr_list;
+  unsigned i = 0;
```
Please separate the declaration from the assignment.

whitlockjc commented on docs/src/fs.rst:
```diff
@@ -301,6 +301,15 @@ API
+.. c:function:: int uv_fs_futime_ex(uv_loop_t* loop, uv_fs_t* req, uv_file file, double btime, double atime, double mtime, uv_fs_cb cb)
+
+    Equivalent to :c:func:`uv_fs_utime` and :c:func:`uv_fs_futime` unless you
+    are on OS X or Windows in which case these variants allow you to also set
```
Should we mention both?  I mean, _"OS X"_ valid prior to the last couple releases _(not sure which it changed off the top of my head)_.  I could see: _"macOS/OS X"_

cjihrig replied:
I think "macOS" has been the way to go since mid 2016 - http://fortune.com/2016/06/13/apple-mac-os-sierra/. I don't have strong opinions on this though.

whitlockjc replied:
Works for me.  Thanks for the quick response.

whitlockjc commented on test/test-fs.c:
```diff
@@ -2199,6 +2218,57 @@ TEST_IMPL(fs_utime) {
+  check_utime_ex(path, btime, atime, mtime);
+  uv_fs_req_cleanup(&req);
+
+  btime = atime = mtime = 1291404900; /* 2010-12-03 20:35:00 - mees <3 */
```
No idea, pre-port.  Let me remove it.  :)

cjihrig replied:
Heh, this was in there a few times.

cjihrig commented on src/win/fs.c:
```diff
@@ -1497,13 +1500,27 @@ static void fs__fchmod(uv_fs_t* req) {
+  FILETIME filetime_m;
+
+  if (isnan(btime)) {
+    &filetime_b = NULL;
```
Wouldn't this just be `filetime_b = NULL;`?

whitlockjc replied:
Copy/paste fail, my bad.

whitlockjc commented:
PR updated with last round of review feedback changes.

cjihrig commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/791/

cjihrig commented:
Looks like all of the failures so far are related except for linux (two of the centos machines haven't started yet).

whitlockjc commented:
I'll look into them this evening.

whitlockjc commented:
The CI page returns a `404` now.  Can we create a new one?  Also...I'm really at a loss on BSD.  I can understand Windows, working on it now, but for BSD...no idea right now.

whitlockjc commented:
Update.  I _think_ that `uv_fs_req_cleanup` is not properly cleaning up `req->statbuf` because even after calling `uv_fs_req_cleanup`, `req.statbuf` still has the previous content in it.  I could be mistaken but my testing seems to suggest this is the case.  I'll test in the morning.

santigimeno commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/815/

whitlockjc commented:
The latest commit updates `fs_futime_ex` and `fs_utime_ex` tests in `test/test-fs.c` to be updated for changes to made to the `fs_futime` and `fs_utime` tests respectively on `master`.  I also updated `test/test-fs.c` to no longer use `math.h` because on AIX/ZOS/... `NAN` and/or `isnan` don't seem to be defined.

I'm working on the failing Windows tests now because I'm at a loss on the failing BSD tests.  Can we get a new CI?

whitlockjc commented:
If someone wants to join me in IRC to work on the BSD stuff, please do.  I'm not sure what's up.  Might even be a test issue.

whitlockjc commented:
In `check_utime_ex` _(`test/test-fs.c`)_, if `s = req.ptr` is used instead of `s = &req->statbuf`, the tests pass on all tested *nix boxes.  I really wonder if `statbuf` is being cleaned up properly via `uv_fs_req_cleanup`.

whitlockjc commented:
@bnoordhuis @saghul @cjihrig I'm really struggling with the BSD failures.  It _seems_ like `&req.statbuf` is retaining state even after `uv_fs_req_cleanup` is called and I _think_ that is why things are failing where the are _(for *nix)_.  I really could use some extra eyes on this.  In the meantime, I'm moving on to Windows failures.

whitlockjc commented:
I'm not sure the Windows errors are related, or fail prior to tests running from what I can see.  I am at a point where I'm not sure what to do next and could really use some help.

whitlockjc commented:
Last commit fixes Windows compilation issues.  Unfortunately, related Windows tests are failing similarly to other failing *nix boxes.  It seems like `uv_fs_req_cleanup` isn't cleaning up all data.  But while on *nix the `req` data ends up corrupted and fails the test assertion, Windows the test process terminates.  I'll keep digging.

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -62,6 +63,43 @@
+    times[*size].tv_sec = req->btime;
+    times[*size].tv_nsec = (unsigned long)(req->btime * 1000000) % 1000000 * 1000;
+
+    (*size)++;
```
`++*size`?  Saves two parens.

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -62,6 +63,43 @@
+
+void uv__prepare_setattrlist_args(uv_fs_t* req,
+                                  struct attrlist* attr_list,
+                                  struct timespec* times,
```
If you specify this as `struct timespec (*times)[3]` and pass it as `&times` at the call site, then callers can't pass in a wrong-sized array.

Aside: this function should be `static`.

Aside 2: I'd `*size = 0` first so callers don't have to (and can't forget.)

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -230,9 +268,17 @@ static ssize_t uv__fs_futime(uv_fs_t* req) {
 
-#elif defined(__APPLE__)                                                      \
-    || defined(__DragonFly__)                                                 \
+  return fsetattrlist(req->file, &attr_list, times, i * sizeof(struct timespec), 0);
```
`sizeof(times[0])`, likewise on line 762.

bnoordhuis commented on test/test-fs.c:
```diff
@@ -2180,7 +2198,7 @@ TEST_IMPL(fs_utime) {
   uv_fs_req_cleanup(&req);
 
-  atime = mtime = 1291404900; /* 2010-12-03 20:35:00 - mees <3 */
+  atime = mtime = 1291404900; /* 2010-12-03 20:35:00 */
```
Don't remove that!  Neither on line 2295, please. :-)

cjihrig replied:
@bnoordhuis just out of curiosity, what is the significance of it?

bnoordhuis replied:
That's when my daughter Mees was born. :-)

bnoordhuis commented on test/test-fs.c:
```diff
@@ -2199,6 +2217,73 @@ TEST_IMPL(fs_utime) {
 
+TEST_IMPL(fs_utime_ex) {
+  utime_check_t checkme;
+  const char* path = "test_file";
```
I'd write this as `const char path[] = "test_file"` since you're not reassigning it, likewise on line 2406.

whitlockjc commented:
@bnoordhuis I've addressed all of your feedback except the one about specifying `struct timespec (*times)[3]` for `uv__prepare_setattrlist_args`.  I couldn't get it to work and I ran out of time.  I'll give it a peek again shortly.

whitlockjc commented:
Remaining feedback from @bnoordhuis has been addressed.  Can we get a new CI?  I don't see it fixing the BSD tests or Windows tests but it would be good to see the new state of things.  I really need some help with the BSD stuff if anyone can help.

cjihrig commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/819/

whitlockjc commented:
@cjihrig Can you jump on #libuv on Freenode really quick?  I'd love to run something by you.

whitlockjc commented:
It seems like `uv_fs_req_cleanup` is causing the Windows test failures after the last update.  I'm working on identifying the cause.

whitlockjc commented:
I think a new CI build _should_ show that Windows tests are fixed.  I'm still stumped by the FreeBSD 10 failure but I've tried the same solutions I used for Windows _(not using `uv__free` when `req->bufs` is empty/uninitialized in `uv_fs_req_cleanup`, etc.)_ but to no avail.  I'm seriously not sure what to do next.

santigimeno commented:
@whitlockjc I've been looking into the `FreeBSD` issue. The issue seems to be that the test is actually setting the birth time. The relevant text from the [utimes(2) man page](https://www.freebsd.org/cgi/man.cgi?query=utimes&sektion=2&manpath=freebsd-release-ports):

> If	times is non-NULL, it is assumed to point to an	array of two timeval
     structures.  The access time is set to the	value of the first element,
     and the modification time is set to the value of the second element.  For
     file systems that support file birth (creation) times (such as UFS2), the
     birth time	will be	set to the value of the	second element if the second
     element is	older than the currently set birth time.  To set both a	birth
     time and a	modification time, two calls are required; the first to	set
     the birth time and	the second to set the (presumably newer) modification
     time.  Ideally a new system call will be added that allows	the setting of
     all three times at	once.  The caller must be the owner of the file	or be
     the super-user.

As the time uses is older than the current set birth time, the birth time is indeed changed.

whitlockjc commented:
I wonder why it only happens on FreeBSD 10 and not FreeBSD 11.  Now that I've seen this, I'll update the tests.

santigimeno commented:
It might be related to the filesystem. On my `FreeBSD 11.1` vm it's happening.

whitlockjc commented:
That makes sense.

whitlockjc commented:
The biggest issue we have now is identifying if/when we should check for equality vs. inequality.  I could just not do the test for FreeBSD when `btime` is not provided to `check_utime_ex`?

/cc @bnoordhuis @cjihrig @saghul @santigimeno

whitlockjc commented:
PR updated with first attempt at handling FreeBSD situation mentioned above.  Can we get a new CI?

santigimeno commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/823/

santigimeno commented:
> I wonder why it only happens on FreeBSD 10 and not FreeBSD 11

I'm curious, do you have a FreeBSD 11 where birth time is not supported? If that's the case could you check the `st_birthtim` field returned by a stat() call?

whitlockjc commented:
So you're saying that instead of making `check_utime_ex` always skew `btime` for FreeBSD, in the caller of `check_utime_ex` we would use `uv_fs_stat` and set the provided `btime` accordingly based on whether or not it changed?  What if we were to just only check `btime` equality when there is a provided `btime` because it seems like certain OSes update `btime` via `utime` and that's not something libuv should be concerned with?  The more I think about that, the more I like it.

whitlockjc commented:
I've updated `check_utime_ex` to no longer check that `btime` is unaltered when the provided `btime` is `-1.0`.  I've documented this in the surrounding code.  At this time, I think this is the best we can do since `utime` _(or underlying OS call)_ cannot be guaranteed to not alter `btime` when none is provided, like on FreeBSD.

The last CI build didn't fail on FreeBSD related to this commit and Windows tests are passing now as well.  Can I get a re-review and new CI?

@bnoordhuis @cjihrig @saghul @santigimeno

refack commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/827/

whitlockjc commented:
It seems like zOS is running into similar issues as other *nix where `math.h` doesn't include `NAN` on all OS versions.  Would anyone be opposed to me removing `math.h` from `src/fs.c` and using `-1` for `NAN` like we did in `test/test-fs.c` during recent PR changes?

Also, OS X seems like it's failing due to an unrelated change.

whitlockjc commented:
The more I think about the `math.h` change, I don't like it because it affects the C API backward compatibility of `uv_fs_futime` and `uv_fs_utime`.

whitlockjc commented:
That being said, I think this PR is now solid.  I'll let y'all decide.

richardlau commented:
> It seems like zOS is running into similar issues as other *nix where math.h doesn't include NAN on all OS versions. 

https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.3.0/com.ibm.zos.v2r3.bpxbd00/mathh.htm says:

> The floating point macros and the macros INFINITY and NAN are supported under IEEE only.

cc @jbarz

santigimeno commented:
> So you're saying that instead of making check_utime_ex always skew btime for FreeBSD, in the caller of check_utime_ex we would use uv_fs_stat and set the provided btime accordingly based on whether or not it changed? What if we were to just only check btime equality when there is a provided btime because it seems like certain OSes update btime via utime and that's not something libuv should be concerned with?

TBH I hadn't gone so far, I was just curious what was the returned value.

> Would anyone be opposed to me removing math.h from src/fs.c and using -1 for NAN like we did in test/test-fs.c during recent PR changes?

Aren't negative integers valid `time_t` values to represent dates before January 1st, 1970?

whitlockjc commented:
Since this is only used for testing, I don't see it being a big deal.  But since we also need to solve the `math.h` stuff for zOS, we could just add it back for tests and fix across the board.

whitlockjc commented:
I've updated tests to use `math.h` again because fixing it is required for runtime anyways.

whitlockjc commented:
Can we get a new CI so we know which OSes need to be fixed with regard to `math.h` compatibility?

santigimeno commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/833/

whitlockjc commented:
It looks like we could add `-Wl,FLOAT` to the zOS compiler options and fix this.  I _think_ we'd do this in `uv.gyp` and `test/test.gyp`.  I don't have a zOS box to test on or I'd just try it.

jBarz commented:
Regarding z/OS, it seems like the merge in f94282c7abc613c388a4172d8b121a3ccc1adbf5 reverted 5d336bd63f06931f54137671dad30abfe12b91a6.
So if the below change from 5d336bd63f06931f54137671dad30abfe12b91a6 can be re-applied, that should find the NAN macro.

```diff
--- a/uv.gyp
+++ b/uv.gyp
@@ -28,7 +28,7 @@
               '_AE_BIMODAL',
               'PATH_MAX=255'
             ],
-            'cflags': [ '-qxplink' ],
+            'cflags': [ '-qxplink', '-qlanglvl=extc99' ],
             'ldflags': [ '-qxplink' ],
           }]
         ],
```

jBarz commented:
Sorry, there appear to be a few more z/OS changes missing. I will have to look at fixing a few of them first. Then you could rebase on top of that. Will keep you informed.

santigimeno commented:
> Regarding z/OS, it seems like the merge in f94282c reverted 5d336bd

Sorry about that

jBarz commented:
No problem at all! I would be ok for this PR to be pushed through regardless.
I will submit(asap) a new PR to get z/OS passing on this master branch.

whitlockjc commented:
@jorangreef I think that would be better as a new PR.  Thoughts?

@jBarz Thanks!

whitlockjc commented:
@jBarz If you don't mind, I can help you avoid the PR and just do it in this PR.

santigimeno commented:
@whitlockjc https://github.com/libuv/libuv/pull/1839 has landed. Would you rebase to current master and will see how the CI goes. Thanks!

whitlockjc commented:
@santigimeno Done.  _(This update **DOES NOT** include the changes suggested by @jBarz.  It seems the delta between now and when zOS changes were made is quite large and I didn't have enough expertise to know the suggested changes were appropriate and/or complete.)_

Can we get a new CI?

santigimeno commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/838/

whitlockjc commented:
Build looks good.  Failing FreeBSD _(`hrtime`)_ should not be as a result of this change and Linux is only on Alpine and it seems to be a build environment issue.

cjihrig commented on src/win/fs.c:
```diff
@@ -1964,7 +1988,8 @@ void uv_fs_req_cleanup(uv_fs_t* req) {
 
-  if (req->fs.info.bufs != req->fs.info.bufsml)
+  if (req->fs.info.bufs != req->fs.info.bufsml
+      && ARRAY_SIZE(req->fs.info.bufs) > 0)
```
Style nit: The `&&` should be on the previous line.

whitlockjc replied:
One example of where this convention is used elsewhere: https://github.com/libuv/libuv/blob/v1.x/src/fs-poll.c#L229

whitlockjc replied:
I went ahead and address this nit, I guess I chose the wrong example to follow.

cjihrig commented on test/test-fs.c:
```diff
@@ -687,24 +689,48 @@ TEST_IMPL(fs_file_loop) {
+  if (!isnan(btime)) {
+    /* Make sure the birth/creation time was altered as expected. */
+    ASSERT(s->st_birthtim.tv_sec + (s->st_birthtim.tv_nsec / 1000000000.0)
+      == btime);
```
Style nit: `==` should go on the previous line.

whitlockjc replied:
I saw this style in other places and followed, same as above.  I can fix it but I'm pretty sure I followed existing convention.

whitlockjc replied:
I went ahead and address this nit, I guess I chose the wrong example to follow.

bnoordhuis commented:
The freebsd CI failure is unrelated and the alpine failures are infrastructural (no automake.)

santigimeno commented:
And finally landed in https://github.com/libuv/libuv/commit/c51ead5c704cae957b9323feef97e842e1bdb033. Thanks!!

jorangreef commented:
Thanks @whitlockjc!

whitlockjc commented:
WOOHOO!!!  I'll get started on the node work now.
